osu.Game/Screens/Select/BeatmapCarousel.cs

```diff
@@ -510,12 +510,27 @@ public bool SelectBeatmap(BeatmapInfo? beatmapInfo, bool bypassFilters = true)
             if (beatmapInfo?.Hidden != false)
                 return false;
 
-            foreach (CarouselBeatmapSet set in beatmapSets)
+            foreach (var carouselItem in root.Items)
             {
+                if (carouselItem is not CarouselBeatmapSet set)
+                    continue;
+
                 if (!bypassFilters && set.Filtered.Value)
                     continue;
 
-                var item = set.Beatmaps.FirstOrDefault(p => p.BeatmapInfo.Equals(beatmapInfo));
+                CarouselBeatmap? item = null;
+
+                foreach (var setCarouselItem in set.Items)
```

Comment:
In addition, this could have still been using `FirstOrDefault` without allocating.